### ISSUE TITLE
Nodes should include start and end for compatibility with espree

### ISSUE THREAD

**Issue by epmatsw** [typescript-eslint/typescript-eslint#405]
**What code were you trying to parse?**

```ts
const blah = () => 'asdf';
```

https://astexplorer.net/#/gist/d48c5e58349ca3fec1d38e1b7a2b6095/5219e84df357a9487224d0b8de72df1adb312cdf

**What did you expect to happen?**

Using `espree` as the parser, the `ArrowFunctionExpression` has both `start` and `end` properties. Same for `babel-eslint`. I would expect `@typescript-eslint/parser` to provide the same API. 

**What actually happened?**

https://astexplorer.net/#/gist/7cc229fb58a8c734eabee9c81b2f63c6/15b0bdef95cacc32dd7c55c42ce54135a02d3ce7

The `ArrowFunctionExpression` doesn't have `start` or `end`.

**Versions**

(Note: I ran into this locally on my machine, but it reproduces easily in astexplorer as linked above. The versions below are from my machine)

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `1.6.0` |
| `TypeScript`                           | `3.4.1` |
| `node`                                 | `11.12.0` |
| `npm`                                  | `6.1.0` |

**Comment by epmatsw** [comment]
The impact of this is that some plugins/rules that rely on `start`/`end` being present (`eslint-plugin-mocha/no-mocha-arrows` for example) will not work when using this parser.

**Comment by bradzacher** [comment]
I'm pretty sure that the `start`/`end` properties on a node are actually non standard.

As per the ESLint docs for writing a custom parser: https://eslint.org/docs/developer-guide/working-with-custom-parsers#all-nodes
The only two required properties on a node are `range` and `loc`.

Technically this is a broken implementation in `eslint-plugin-mocha`, as they are using non-standard properties to implement their rule.
Looking at the source code, that would be correct, as they are passing `start`/`end` into fixer functions, when they should be using `range[0]`/`range[1]`.

I would consider filing a bug with them, as fixing their rule is much easier than fixing the parser!

---

Not sure how much work it is to add this in, but if there are enough rules that break because of it, we can look into adding it.

This is the first plugin rule that we've had reported broken because of the missing non-standard properties though.

**Comment by sindresorhus** [comment]
Relevant ESLint commit: https://github.com/eslint/eslint/commit/b3e4598

Instead of implementing those properties here, I would rather see an issue on ESLint about hard-deprecating those properties in their next major version.

// @mysticatea

**Comment by sindresorhus** [comment]
The main reason many rules use `start`/`end` is that https://astexplorer.net/ display them when selecting the `espree` parser. Someone should open an issue here too: https://github.com/fkling/astexplorer/issues

**Comment by sindresorhus** [comment]
`start`/`end` doesn't even always return the same values as `node.range[0]`/`node.range[1]`...

If you click inside the string in https://astexplorer.net/#/gist/f31bebfae0fa9a0cc7d0ec88f6a0daa6/94ef9347110394fe2256ea8406e24fbb9fa58e31, you'll see `start: 13, end: 20, range: [12, 21]`.

**Comment by bradzacher** [comment]
Yeah that's what I noticed too - I was comparing them but noticed that the end was usually off by one. I didn't see the start was wrong, but that makes it worse!
There must be some weird interactions when rules use those in their fixers.

I'm going to close this issue as we're not going to add the deprecated/dead/non-standard properties into the parser.

Please file an issue/pr with the specified plugin to fix their rule.

**Comment by mysticatea** [comment]
I took a memo about adding the check that prevents the use of `node.start`/`node.end` into `RuleTester`: https://github.com/eslint/eslint/projects/6
(it's going to be on 7.0.0 line because 6.0.0 has been feature-freezed already)
